File: frontend/src/constants/takt-constants.ts
// ========================================
// 项目名称：节拍工厂·Takt Plat
// 命名空间：frontend/src/constants
// 文件名称：takt-constants.ts
// 创建时间：2026-06-25
// 创建人：Takt365(Cursor AI)
// 功能描述：系统内置常量与展示文案（与 backend TaktConstants 对齐；非字典项）
//
// 版权信息：Copyright (c) 2025 Takt  All rights reserved.
// 免责声明：此软件使用 MIT License，作者不承担任何使用风险。
// ========================================

/** 内置常量类别 */
export type TaktConstCategory =
  | 'loginType'
  | 'loginResult'
  | 'operType'
  | 'quartzTaskType'
  | 'deviceType'
  | 'browserType'
  | 'operatingSystem'

/** 登录方式（TaktConstants.LoginType） */
export const TaktLoginType = {
  Unknown: 'unknown',
  Password: 'password',
  RefreshToken: 'refreshtoken',
  ClientCredentials: 'clientcredentials',
  AuthorizationCode: 'authorizationcode',
  OAuthAuthorize: 'oauthauthorize',
  VerifyPassword: 'verifypassword',
  SignOut: 'signout',
} as const

/** 登录结果（TaktConstants.LoginResult） */
export const TaktLoginResult = {
  Success: 'success',
  PasswordError: 'passworderror',
  UserNotFound: 'usernotfound',
  UserDisabled: 'userdisabled',
  UserLocked: 'userlocked',
  CaptchaError: 'captchaerror',
} as const

/** 操作类型（TaktConstants.OperType） */
export const TaktOperType = {
  Unknown: 'unknown',
  Create: 'create',
  Update: 'update',
  Delete: 'delete',
  Query: 'query',
  Export: 'export',
  Import: 'import',
  Grant: 'grant',
  ForceOut: 'forceout',
  CodeGen: 'codegen',
  ClearData: 'cleardata',
} as const

/** Quartz 任务类型（TaktConstants.QuartzTaskType） */
export const TaktQuartzTaskType = {
  Assembly: 'assembly',
  Http: 'http',
  Sql: 'sql',
} as const

/** Quartz 触发器类型（对齐 TaktQuartzSchedulerManager：0=Simple 1=Cron） */
export const TaktQuartzTriggerType = {
  Simple: 0,
  Cron: 1,
} as const

/** Quartz Misfire 策略（对齐 TaktQuartzSchedulerManager：0=默认 1=忽略 2=立即触发 3=不触发） */
export const TaktQuartzMisfirePolicy = {
  Default: 0,
  Ignore: 1,
  FireAndProceed: 2,
  DoNothing: 3,
} as const

/** 登录设备（TaktConstants.DeviceType） */
export const TaktDeviceType = {
  Unknown: 'unknown',
  Pc: 'pc',
  Mobile: 'mobile',
  Tablet: 'tablet',
} as const

/** 浏览器（TaktConstants.BrowserType） */
export const TaktBrowserType = {
  Unknown: 'unknown',
  Chrome: 'chrome',
  Firefox: 'firefox',
  Safari: 'safari',
  Edge: 'edge',
} as const

/** 操作系统（TaktConstants.OperatingSystem） */
export const TaktOperatingSystem = {
  Unknown: 'unknown',
  Windows: 'windows',
  MacOs: 'macos',
  Linux: 'linux',
  Android: 'android',
  Ios: 'ios',
} as const

const LABELS: Record<TaktConstCategory, Record<string, string>> = {
  loginType: {
    [TaktLoginType.Unknown]: '未知',
    [TaktLoginType.Password]: '账号密码',
    [TaktLoginType.RefreshToken]: '刷新令牌',
    [TaktLoginType.ClientCredentials]: '客户端凭证',
    [TaktLoginType.AuthorizationCode]: '授权码换令牌',
    [TaktLoginType.OAuthAuthorize]: 'OAuth授权登录',
    [TaktLoginType.VerifyPassword]: '预检验密',
    [TaktLoginType.SignOut]: '注销会话',
  },
  loginResult: {
    [TaktLoginResult.Success]: '登录成功',
    [TaktLoginResult.PasswordError]: '密码错误',
    [TaktLoginResult.UserNotFound]: '用户不存在',
    [TaktLoginResult.UserDisabled]: '用户已禁用',
    [TaktLoginResult.UserLocked]: '用户已锁定',
    [TaktLoginResult.CaptchaError]: '验证码错误',
  },
  operType: {
    [TaktOperType.Unknown]: '未知',
    [TaktOperType.Create]: '新增',
    [TaktOperType.Update]: '修改',
    [TaktOperType.Delete]: '删除',
    [TaktOperType.Query]: '查询',
    [TaktOperType.Export]: '导出',
    [TaktOperType.Import]: '导入',
    [TaktOperType.Grant]: '授权',
    [TaktOperType.ForceOut]: '强退',
    [TaktOperType.CodeGen]: '生成代码',
    [TaktOperType.ClearData]: '清空数据',
  },
  quartzTaskType: {
    [TaktQuartzTaskType.Assembly]: '程序集',
    [TaktQuartzTaskType.Http]: '网络请求',
    [TaktQuartzTaskType.Sql]: 'SQL语句',
  },
  deviceType: {
    [TaktDeviceType.Unknown]: '未知',
    [TaktDeviceType.Pc]: 'PC',
    [TaktDeviceType.Mobile]: '手机',
    [TaktDeviceType.Tablet]: '平板',
  },
  browserType: {
    [TaktBrowserType.Unknown]: '未知',
    [TaktBrowserType.Chrome]: 'Chrome',
    [TaktBrowserType.Firefox]: 'Firefox',
    [TaktBrowserType.Safari]: 'Safari',
    [TaktBrowserType.Edge]: 'Edge',
  },
  operatingSystem: {
    [TaktOperatingSystem.Unknown]: '未知',
    [TaktOperatingSystem.Windows]: 'Windows',
    [TaktOperatingSystem.MacOs]: 'macOS',
    [TaktOperatingSystem.Linux]: 'Linux',
    [TaktOperatingSystem.Android]: 'Android',
    [TaktOperatingSystem.Ios]: 'iOS',
  },
}

/**
 * 解析内置常量展示文案
 * @param category 常量类别
 * @param value 落库值
 * @returns 展示文案；未知则回退原值或「-」
 */
export function formatTaktConstLabel(
  category: TaktConstCategory,
  value?: string | null,
): string {
  if (value == null || value === '') {
    return '-'
  }
  const trimmed = String(value).trim()
  return LABELS[category][trimmed] ?? trimmed
}

/**
 * 内置常量下拉选项
 * @param category 常量类别
 * @returns label/value 选项列表
 */
export function getTaktConstOptions(category: TaktConstCategory): Array<{ label: string; value: string }> {
  const map = LABELS[category]
  return Object.entries(map).map(([value, label]) => ({ label, value }))
}

export const loginTypeOptions = getTaktConstOptions('loginType')
export const loginResultOptions = getTaktConstOptions('loginResult')
export const operTypeOptions = getTaktConstOptions('operType')
export const quartzTaskTypeOptions = getTaktConstOptions('quartzTaskType')
export const browserTypeOptions = getTaktConstOptions('browserType')
export const operatingSystemOptions = getTaktConstOptions('operatingSystem')

File: backend/src/Takt.Shared/Constants/TaktConstants.cs
// ========================================
// 项目名称：节拍工厂·Takt Plat
// 命名空间：Takt.Shared.Constants
// 文件名称：TaktConstants.cs
// 创建时间：2026-06-25
// 创建人：Takt365(Cursor AI)
// 功能描述：系统内置常量（落库字段值权威来源；非字典项，前后端对齐 TaktConstants）
//
// 版权信息：Copyright (c) 2025 Takt  All rights reserved.
// 免责声明：此软件使用 MIT License，作者不承担任何使用风险。
// ========================================

namespace Takt.Shared.Constants;

/// <summary>
/// 系统内置常量（登录/操作/Quartz/设备等；写库与校验唯一引用点）
/// </summary>
public static class TaktConstants
{
    /// <summary>
    /// 登录方式
    /// </summary>
    public static class LoginType
    {
        /// <summary>
        /// 未知
        /// </summary>
        public const string Unknown = "unknown";
        /// <summary>
        /// 账号密码
        /// </summary>
        public const string Password = "password";
        /// <summary>
        /// 刷新令牌
        /// </summary>
        public const string RefreshToken = "refreshtoken";
        /// <summary>
        /// 客户端凭证
        /// </summary>
        public const string ClientCredentials = "clientcredentials";
        /// <summary>
        /// 授权码换令牌
        /// </summary>
        public const string AuthorizationCode = "authorizationcode";
        /// <summary>
        /// OAuth 授权页登录
        /// </summary>
        public const string OAuthAuthorize = "oauthauthorize";
        /// <summary>
        /// 登录预检验密
        /// </summary>
        public const string VerifyPassword = "verifypassword";
        /// <summary>
        /// 注销会话
        /// </summary>
        public const string SignOut = "signout";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Unknown, Password, RefreshToken, ClientCredentials, AuthorizationCode, OAuthAuthorize, VerifyPassword, SignOut,
        };

        /// <summary>
        /// 是否为允许的登录方式常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// 登录结果
    /// </summary>
    public static class LoginResult
    {
        /// <summary>
        /// 登录成功
        /// </summary>
        public const string Success = "success";
        /// <summary>
        /// 密码错误
        /// </summary>
        public const string PasswordError = "passworderror";
        /// <summary>
        /// 用户不存在
        /// </summary>
        public const string UserNotFound = "usernotfound";
        /// <summary>
        /// 用户已禁用
        /// </summary>
        public const string UserDisabled = "userdisabled";
        /// <summary>
        /// 用户已锁定
        /// </summary>
        public const string UserLocked = "userlocked";
        /// <summary>
        /// 验证码错误
        /// </summary>
        public const string CaptchaError = "captchaerror";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Success, PasswordError, UserNotFound, UserDisabled, UserLocked, CaptchaError,
        };

        /// <summary>
        /// 是否为允许的登录结果常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// 操作类型
    /// </summary>
    public static class OperType
    {
        /// <summary>
        /// 未知
        /// </summary>
        public const string Unknown = "unknown";
        /// <summary>
        /// 新增
        /// </summary>
        public const string Create = "create";
        /// <summary>
        /// 修改
        /// </summary>
        public const string Update = "update";
        /// <summary>
        /// 删除
        /// </summary>
        public const string Delete = "delete";
        /// <summary>
        /// 查询
        /// </summary>
        public const string Query = "query";
        /// <summary>
        /// 导出
        /// </summary>
        public const string Export = "export";
        /// <summary>
        /// 导入
        /// </summary>
        public const string Import = "import";
        /// <summary>
        /// 授权
        /// </summary>
        public const string Grant = "grant";
        /// <summary>
        /// 强退
        /// </summary>
        public const string ForceOut = "forceout";
        /// <summary>
        /// 生成代码
        /// </summary>
        public const string CodeGen = "codegen";
        /// <summary>
        /// 清空数据
        /// </summary>
        public const string ClearData = "cleardata";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Unknown, Create, Update, Delete, Query, Export, Import, Grant, ForceOut, CodeGen, ClearData,
        };

        /// <summary>
        /// 是否为允许的操作类型常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// Quartz 任务类型
    /// </summary>
    public static class QuartzTaskType
    {
        /// <summary>
        /// 程序集
        /// </summary>
        public const string Assembly = "assembly";
        /// <summary>
        /// 网络请求
        /// </summary>
        public const string Http = "http";
        /// <summary>
        /// SQL 语句
        /// </summary>
        public const string Sql = "sql";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Assembly, Http, Sql,
        };

        /// <summary>
        /// 是否为允许的任务类型常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// Quartz 触发器类型（对齐 TaktQuartzSchedulerManager：0=Simple 1=Cron）
    /// </summary>
    public static class QuartzTriggerType
    {
        /// <summary>
        /// Simple 间隔触发
        /// </summary>
        public const int Simple = 0;
        /// <summary>
        /// Cron 表达式触发
        /// </summary>
        public const int Cron = 1;

        private static readonly HashSet<int> Allowed = new()
        {
            Simple, Cron,
        };

        /// <summary>
        /// 是否为允许的触发器类型
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(int value) => Allowed.Contains(value);
    }

    /// <summary>
    /// Quartz Misfire 策略（对齐 TaktQuartzSchedulerManager：0=默认 1=忽略 2=立即触发 3=不触发）
    /// </summary>
    public static class QuartzMisfirePolicy
    {
        /// <summary>
        /// 默认策略
        /// </summary>
        public const int Default = 0;
        /// <summary>
        /// 忽略 Misfire
        /// </summary>
        public const int Ignore = 1;
        /// <summary>
        /// 立即触发
        /// </summary>
        public const int FireAndProceed = 2;
        /// <summary>
        /// 不触发
        /// </summary>
        public const int DoNothing = 3;

        private static readonly HashSet<int> Allowed = new()
        {
            Default, Ignore, FireAndProceed, DoNothing,
        };

        /// <summary>
        /// 是否为允许的 Misfire 策略
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(int value) => Allowed.Contains(value);
    }

    /// <summary>
    /// 登录设备
    /// </summary>
    public static class DeviceType
    {
        /// <summary>
        /// 未知
        /// </summary>
        public const string Unknown = "unknown";
        /// <summary>
        /// PC
        /// </summary>
        public const string Pc = "pc";
        /// <summary>
        /// 手机
        /// </summary>
        public const string Mobile = "mobile";
        /// <summary>
        /// 平板
        /// </summary>
        public const string Tablet = "tablet";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Unknown, Pc, Mobile, Tablet,
        };

        /// <summary>
        /// 是否为允许的登录设备常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// 浏览器
    /// </summary>
    public static class BrowserType
    {
        /// <summary>
        /// 未知
        /// </summary>
        public const string Unknown = "unknown";
        /// <summary>
        /// Chrome
        /// </summary>
        public const string Chrome = "chrome";
        /// <summary>
        /// Firefox
        /// </summary>
        public const string Firefox = "firefox";
        /// <summary>
        /// Safari
        /// </summary>
        public const string Safari = "safari";
        /// <summary>
        /// Edge
        /// </summary>
        public const string Edge = "edge";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Unknown, Chrome, Firefox, Safari, Edge,
        };

        /// <summary>
        /// 是否为允许的浏览器常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }

    /// <summary>
    /// 审计日志操作人登录名（无法解析时落库值）
    /// </summary>
    public static class AuditUserName
    {
        /// <summary>
        /// 未知操作人
        /// </summary>
        public const string Unknown = "unknown";
    }

    /// <summary>
    /// 无登录上下文时的审计操作人 ID（与种子 admin 工号 900001、菜单种子 CreatedBy 一致）
    /// </summary>
    public static class SystemAuditUser
    {
        /// <summary>
        /// 系统默认审计用户 ID
        /// </summary>
        public const long Id = 900001L;
    }

    /// <summary>
    /// 操作系统
    /// </summary>
    public static class OperatingSystem
    {
        /// <summary>
        /// 未知
        /// </summary>
        public const string Unknown = "unknown";
        /// <summary>
        /// Windows
        /// </summary>
        public const string Windows = "windows";
        /// <summary>
        /// macOS
        /// </summary>
        public const string MacOs = "macos";
        /// <summary>
        /// Linux
        /// </summary>
        public const string Linux = "linux";
        /// <summary>
        /// Android
        /// </summary>
        public const string Android = "android";
        /// <summary>
        /// iOS
        /// </summary>
        public const string Ios = "ios";

        private static readonly HashSet<string> Allowed = new(StringComparer.Ordinal)
        {
            Unknown, Windows, MacOs, Linux, Android, Ios,
        };

        /// <summary>
        /// 是否为允许的操作系统常量值
        /// </summary>
        /// <param name="value">待校验值</param>
        /// <returns>是否允许</returns>
        public static bool IsValid(string? value)
        {
            return !string.IsNullOrWhiteSpace(value) && Allowed.Contains(value.Trim());
        }
    }
}
